### PR TITLE
refactor: convert `*like_dyn`, `*like_utf8_scalar_dyn` and  `*like_dict` functions to macros

### PR DESCRIPTION
**Pull request by askoa**
# Which issue does this PR close?

Part of #3296 

# What changes are included in this PR?
Three declarative macro added to generate 12 functions below. 
  -  `*like_dyn` functions: `like_dyn`, `nlike_dyn`, `ilike_dyn` and `nilike_dyn`.
  -  `*like_utf8_scalar_dyn` functions: `like_utf8_scalar_dyn`, `nlike_utf8_scalar_dyn`, `ilike_utf8_scalar_dyn` and `nilike_utf8_scalar_dyn`.
  -  `*like_dict` functions: `like_dict`, `nlike_dict`, `ilike_dict` and `nilike_dict`.

# Are there any user-facing changes?
No.

### CODE REVIEW

File: arrow-string/src/like.rs [hunk 1]
```diff
@@ -25,6 +25,135 @@ use arrow_select::take::take;
...
+    }
+}
+static LIKE_DYN: &str = "like_dyn";
+static N_LIKE_DYN: &str = "nlike_dyn";
+static I_LIKE_DYN: &str = "ilike_dyn";
+static N_I_LIKE_DYN: &str = "nilike_dyn";
+static LIKE_FN: &LikeFunction = &LikeFunction {};
+static N_LIKE_FN: &NLikeFunction = &NLikeFunction {};
+static I_LIKE_FN: &ILikeFunction = &ILikeFunction {};
+static N_I_LIKE_FN: &NILikeFunction = &NILikeFunction {};
```
Comment (askoa) [inline arrow-string/src/like.rs:113]: Should I use `lazy_static`? Not sure in what scenarios to use it.

**Comment by askoa** [comment]
No significant changes in benchmarks

```
ilike_utf8 scalar equals
                        time:   [243.40 µs 244.81 µs 246.26 µs]
                        change: [-3.1756% -2.1104% -1.0141%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Benchmarking ilike_utf8 scalar contains: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 9.5s, enable flat sampling, or reduce sample count to 50.
ilike_utf8 scalar contains
                        time:   [1.8345 ms 1.8541 ms 1.8767 ms]
                        change: [-2.4958% -0.9651% +0.5636%] (p = 0.22 > 0.05)
                        No change in performance detected.

ilike_utf8 scalar ends with
                        time:   [245.18 µs 247.24 µs 249.71 µs]
                        change: [+1.6417% +3.0126% +4.4299%] (p = 0.00 < 0.05)
                        Performance has regressed.

ilike_utf8 scalar starts with
                        time:   [270.07 µs 272.10 µs 274.49 µs]
                        change: [-0.7759% +0.0457% +0.8078%] (p = 0.91 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) high mild
  8 (8.00%) high severe

Benchmarking ilike_utf8 scalar complex: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 10.0s, enable flat sampling, or reduce sample count to 40.
ilike_utf8 scalar complex
                        time:   [1.8412 ms 1.8614 ms 1.8841 ms]
                        change: [-3.4178% -2.0494% -0.6609%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 17 outliers among 100 measurements (17.00%)
  3 (3.00%) high mild
  14 (14.00%) high severe

nilike_utf8 scalar equals
                        time:   [248.11 µs 249.37 µs 250.82 µs]
                        change: [-3.1327% -2.3248% -1.5422%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

Benchmarking nilike_utf8 scalar contains: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 9.6s, enable flat sampling, or reduce sample count to 50.
nilike_utf8 scalar contains
                        time:   [1.8652 ms 1.8865 ms 1.9081 ms]
                        change: [-1.0305% +0.3844% +1.8825%] (p = 0.61 > 0.05)
                        No change in performance detected.

nilike_utf8 scalar ends with
                        time:   [293.38 µs 297.60 µs 302.06 µs]
                        change: [+1.2162% +2.5529% +3.8868%] (p = 0.00 < 0.05)
                        Performance has regressed.

nilike_utf8 scalar starts with
                        time:   [277.38 µs 280.41 µs 283.75 µs]
                        change: [-2.1562% -1.4000% -0.5886%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) high mild
  7 (7.00%) high severe

Benchmarking nilike_utf8 scalar complex: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 9.6s, enable flat sampling, or reduce sample count to 50.
nilike_utf8 scalar complex
                        time:   [1.8655 ms 1.8859 ms 1.9085 ms]
                        change: [-1.0484% +0.1324% +1.3700%] (p = 0.83 > 0.05)
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild

ilike_utf8_scalar_dyn dictionary[10] string[4])
                        time:   [83.185 µs 83.797 µs 84.536 µs]
                        change: [-1.7924% -1.1183% -0.4851%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe


```

**Review by tustvold — COMMENTED** [review]
I wonder if this would be simpler as a declarative macro?

**Comment by askoa** [comment]
@tustvold not written a declarative macro so far.  Let me look it up.

File: arrow-string/src/like.rs [hunk 2]
```diff
@@ -25,6 +25,53 @@ use arrow_select::take::take;
 use regex::Regex;
 use std::collections::HashMap;
 
+macro_rules! dyn_function {
+    ($sql:tt, $fn_name:tt, $fn_utf8:tt, $fn_dict:tt) => {
+#[doc = concat!("Perform SQL `left ", $sql ," right` operation on [`StringArray`] /")]
+/// [`LargeStringArray`], or [`DictionaryArray`] with values
+/// [`StringArray`]/[`LargeStringArray`].
+///
+/// See the documentation on [`like_utf8`] for more details.
```
Comment (tustvold) [inline arrow-string/src/like.rs:34]: I think this should be `ilike_utf8` for the ilike kernels. How about documenting the functions outside the macros?

e.g.

```
/// See the documentation on [`like_utf8`] for more details.
dyn_function!(like_dyn, like_utf8, like_dict);
```

**Review by tustvold — APPROVED** [review]
Just a minor documentation nit, looks good thank you :smile:

**Review by viirya — APPROVED** [review]

File: arrow-string/src/like.rs (hunk 2 above)
Comment (askoa) [reply arrow-string/src/like.rs]: The `*like_utf8` name is already captured in the macro. So I just added a document line.  I just don't see the benefits of comment outside the macro vs inside it. So I left the comments inside the macro.

Comment (askoa) [reply arrow-string/src/like.rs:34]: I changed the documentation back to `like_utf8` as I realized it is on purpose. The current code points to `like_utf8` in all functions as its the only function with an example.

**Pull request merged**